jax/lax/lax_parallel.py

```diff
@@ -387,12 +388,19 @@ def _psum_transpose_rule(cts, axis_name, axis_index_groups):
 pxla.parallel_pure_rules[psum_p] = lambda *args, shape: (x * prod(shape) for x in args)
 ad.deflinear(psum_p, _psum_transpose_rule)
 pxla.multi_host_supported_collectives.add(psum_p)
+batching.collective_rules[psum_p] = \
+    lambda vals, dims, axis_size, **_: [v.sum(d) if d is not batching.not_mapped else
```

Comment:
Currently we might need to lower this to a reduce-sum followed by a broadcast, as per [this comment](https://github.com/google/jax/issues/3970#issuecomment-670594807) and [this one](https://github.com/google/jax/issues/3970#issuecomment-670672926). But also we might want to revise that, so that psum just means reduce-sum... I'm not sure what the best thing to do is, just wanted to flag this as a discussion topic!

Reply:
I think that the broadcast is always implicit, because we never wrap results of collectives back in `BatchTracer`s. So they should just get broadcast at each use-site, but you can also return them with `out_axes=None`. Does that make sense?

Reply:
I think that often works but not always when using `grad` inside the `vmap`. That is, unless we broadcast I think (but I am very unsure!) that this answer might be different between `vmap` and `pmap`:

```python
def f(w, data):
  def agg_loss(w):
    return lax.psum(w * data, 'batch')
  return grad(agg_loss)(w)
print(either_vmap_or_pmap(f, axis_name='batch')(jnp.ones(2), jnp.arange(2)))
```

Well, I'd guess guess it'd always be different: if you add a broadcast it may be an error with `vmap` (but work with `pmap`), and without the broadcast it may work with both but produce different answers. That's just guessing though.

We could just land and iterate about this. Within <1mo we may be able to bring them into agreement around reduce-sum (no broadcast) anyway. But it might be worth double checking.

Reply:
Right, good point! I checked the snippet and right now vmap and pmap actually do give the same answer, so I guess we shouldn't block the merge on that. Whether this is the answer we expect it to give is a whole another thing, but we should also make sure that the behave consistently, which they do at the moment.

Reply:
What about
```python
def f(w, data):
  def agg_loss(w):
    return lax.psum(w * data, 'batch')
  return grad(agg_loss)(w)
print(either_vmap_or_pmap(f, axis_name='batch', in_axes=(None, 0))(1., jnp.arange(2)))
```
I think that currently gives a mapped (unsummed) value for pmap, and "implicit broadcasting" should make it give an unmapped (summed) value for vmap.

Reply:
(That seems to be an existing inconsistency between vmap and pmap, not one introduced by collectives.)